File: webmvp/src/lib/usernameSuggestions.ts
import { normalizeUsername } from "@/lib/validation";

/** Suggest a valid username from Google/email profile data. */
export function suggestUsername(
  email?: string | null,
  displayName?: string | null,
): string {
  if (displayName?.trim()) {
    const fromName = normalizeUsername(
      displayName.trim().replace(/\s+/g, "_"),
    ).replace(/[^a-z0-9_]/g, "");
    if (fromName.length >= 3) {
      return fromName.slice(0, 20);
    }
  }

  if (email) {
    const prefix = email.split("@")[0] ?? "";
    const normalized = normalizeUsername(prefix).replace(/[^a-z0-9_]/g, "");
    if (normalized.length >= 3) {
      return normalized.slice(0, 20);
    }
  }

  return "";
}

File: mobile/lib/domain/username_suggestions.dart
import 'package:lf_chords/domain/validation.dart';

/// Suggest a valid username from Google/email profile data.
String suggestUsername(String? email, String? displayName) {
  if (displayName != null && displayName.trim().isNotEmpty) {
    final fromName = normalizeUsername(
      displayName.trim().replaceAll(RegExp(r'\s+'), '_'),
    ).replaceAll(RegExp(r'[^a-z0-9_]'), '');
    if (fromName.length >= 3) {
      return fromName.substring(0, fromName.length.clamp(0, 20));
    }
  }

  if (email != null && email.isNotEmpty) {
    final prefix = email.split('@').first;
    final normalized =
        normalizeUsername(prefix).replaceAll(RegExp(r'[^a-z0-9_]'), '');
    if (normalized.length >= 3) {
      return normalized.substring(0, normalized.length.clamp(0, 20));
    }
  }

  return '';
}
